nova_vm/src/heap/heap_bits.rs

```diff
@@ -1063,3 +1066,30 @@ pub(crate) fn sweep_heap_elements_vector_descriptors<T>(
         descriptors.insert(new_key, descriptor);
     }
 }
+
+#[cfg(feature = "array-buffer")]
+pub(crate) fn sweep_data_view_side_table_values<V: Copy>(
+    // TODO: Figure out a generic way to handle all types which wrap `BaseIndex`
+    side_table: &mut AHashMap<DataView, V>,
+    compaction_list: &CompactionList,
+) {
+    let mut moved_byte_lengths = side_table
+        .iter()
+        .map(|(&data_view, &value)| {
+            let data_view_index = data_view.0.into_u32_index();
+            (
+                data_view_index,
+                data_view_index - compaction_list.get_shift_for_index(data_view_index),
```

Comment:
**issue:** This is unfortunately not quite yet enough :( If memory serves, the shift indexes always point to a live item, eg. "shift down by two from index 15" means that item at index 15 is still live and must become item at index 13 next.

But what about items 13 and 14? They're not necessarily in any shift group but they should be (conceptually) shifted out of the heap entirely, or in other words deleted.

The shift index system is a bit ill-prepared to handle these cases. For the vectors it makes sense to just use the mark booleans to determine if an item should be dropped or not since they're in the proper order just like the vectors themselves.

For strong references the index shifts makes sense as they are, as we never need to worry about shifting a strong reference that points to an item that got removed. We only shift our references if we're alive, and because we're alive our pointed-to strong references must also be alive.

But for weak references we cannot be sure that our target is alive even though we are. The keys here are effectively weak references (in a weird way). I think we might best add a second method to the compaction list for use on weak references that returns `Some(shift)` or `None` to indicate that this item was removed.

Then here you might do a filter_map to return `(old_index, None)` for removed indexes and `(old_index, Some(new_index, value))` for shifted indexes, or something like that.

Reply:
Ok, so would a `get_shift_for_weak_index` (name suggestions?) then essentially be the same as `get_shift_for_index` except for the last `.unwrap_or`?:

```rust
    pub fn get_shift_for_index(&self, index: u32) -> u32 {
        self.indexes
            .iter()
            .enumerate()
            .rev()
            .find(|(_, candidate)| **candidate <= index)
            .map(|(index, _)| *self.shifts.get(index).unwrap())
            .unwrap_or(0)
    }
```

Reply:
This was the solution I wrote:
```
#[cfg(feature = "array-buffer")]
pub(crate) fn sweep_data_view_side_table_values<V: Copy>(
    // TODO: Figure out a generic way to handle all types which wrap `BaseIndex`
    side_table: &mut AHashMap<DataView, V>,
    compaction_list: &CompactionList,
) {
    let mut changed_byte_lengths = side_table
        .iter()
        .filter_map(|(&data_view, &value)| {
            let data_view_index = data_view.0.into_u32_index();
            let Some(shift) = compaction_list.get_shift_for_weak_index(data_view_index) else {
                return Some((data_view_index, None));
            };
            
            if shift == 0 {
                return None;
            };

            println!("Shifting {} by {}", data_view_index, shift);

            Some((data_view_index, Some((data_view_index - shift, value))))
        })
        .collect::<Vec<_>>();

    changed_byte_lengths.sort_by_key(|(old_index, _)| *old_index);

    for (old_index, maybe_moved) in changed_byte_lengths {
        if let Some((new_index, value)) = maybe_moved {
            println!("Moving {} -> {}", old_index, new_index);
            side_table.insert(DataViewIndex::from_u32_index(new_index).into(), value);
            side_table.remove(&DataViewIndex::from_u32_index(old_index).into());
        } else {
            println!("Deleting {}", old_index);
            side_table.remove(&DataViewIndex::from_u32_index(old_index).into());
        }
    }
}
```

Which doesn't really work at all because the shift for index 1 is somehow 2?
```
Shifting 2 by 2
Shifting 1 by 2
```

In other local test cases though I seem to have gotten it to work:
```
Deleting 0
Moving 1 -> 0
Moving 2 -> 1
Deleting 0
Moving 1 -> 0
Moving 2 -> 1
```

Reply:
> Am I on the right track or would a solution look more like the sweep_heap_elements_vector_descriptors function?

Ended up doing this and now I seem to get the expected output from my "tests".